README.md
Reversi Server & Client 
=======================

ビルドの仕方
------------

    $ stack build 

使い方
------

審判サーバを起動：

    $ stack exec reversi-serv -- -p 30000 -t 500


`-p`はポートの指定．
起動すると 

    Waiting 2 connections ...

と表示される．


ここで、クライアントを起動：

    $ stack exec reversi -- -H "localhost" -p 30000 -n Player1 

すると，サーバ側の表示が,

    Waiting 1 connection ... 

となるので,もう一つクライアントを

    $ stack exec reversi -- -H "localhost" -p 30000 -n Player2

のように起動すると対戦が進む．プレイヤー名は各自で異なっている必要あり．
（黒と白を交替しつつ偶数回対戦する）


サーバのオプション
------------------

`-p`, `--port`
: ポート番号（デフォルトは3000）

`-t`, `--playerstime` 
: 各プレイヤーの持ち時間（デフォルトは十分間）

`--timeout`
: コマンドを待つ許容時間．この時間より遅くなるとタイムアウト（デフォルトは1分）

`-T`, `--byoyomi`
: 持ち時間を使い果した後でも使える時間（デフォルトは500 ms）

`-n`
: プレイヤーの人数．サーバはこの人数の接続があると総当たり戦を始める（デフォルトは2）．

`-r`, `--rounds`
: 二人のプレイヤーはこの回数だけ対戦する（デフォルトは4）．

`-c`, `--concurrency`
: 並行で行う対戦の数．n>=4でないと意味を無さない（デフォルトは2）．


クライアントのオプション
------------------------

`-H`
: サーバのホスト名（デフォルトはlocalhost）

`-p`
: サーバのポート番号（デフォルトは3000）

`-n`
: プレイヤー名．現在サーバはこの値を元に結果を集計するので互いに異なる必要がある．


既知の問題点
------------
  * クライアントが異常終了した場合の挙動があやしい


ChangeLog.md
# Revision history for reversi-hs

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


Play.hs
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE MultiWayIf #-}
module Play where

import           Control.Monad
import           Control.Monad.Loops (allM)
import           Data.Array
import           Data.Array.IO
import           Data.Array.MArray
import           Data.Either
import           Data.Foldable       (foldrM)
import           Data.List           (maximumBy, splitAt, (\\))
import           Data.Tuple          (swap)

import           Color
import           Command
import           System.Random


-- 番兵付きの10x10配列
type Board = IOUArray (Int,Int) Int

initBoard :: IO Board
initBoard =
    do board <- newArray ( (0,0), (9,9) ) none
       mapM_ (\i ->
                  do writeArray board (i,0) sentinel
                     writeArray board (i,9) sentinel
                     writeArray board (0,i) sentinel
                     writeArray board (9,i) sentinel) [0..9]
       writeArray board (4,4) white
       writeArray board (4,5) black
       writeArray board (5,4) black
       writeArray board (5,5) white
       return board

isValidMove :: Board -> Color -> (Int,Int) -> IO Bool
isValidMove board color (i,j) =
    do e <- readArray board (i,j)
       if e == none then
           isEffective board color (i,j)
       else
           return False
-- 8方向
dirs = [ (i,j) | i <- [1,0,-1], j <- [1,0,-1] ] \\ [(0,0)]

-- 石を置いたときに、ひっくり返せるかどうか
isEffective :: Board -> Color -> (Int,Int) -> IO Bool
isEffective board color (i,j) =
    do ms <- flippableIndices board color (i,j)
       return $ not $ null ms

-- 石を置いたときにひっくり返えるところ
flippableIndices :: Board -> Color -> (Int,Int) -> IO [(Int,Int)]
flippableIndices board color (i,j) =
    do bs <- mapM (\(di,dj) -> flippableIndicesLine board color (di,dj) (i+di,j+dj)) dirs
       return $ concat bs

flippableIndicesLine board color (di,dj) (i,j) =
    checkLine (di,dj) (i,j) []
    where
      ocolor = oppositeColor color
      checkLine (di,dj) (i,j) r =
          do c <- readArray board (i,j)
             if c == ocolor then
                 checkLine' (di,dj) (i+di,j+dj) ( (i,j) : r )
             else
                 return []
      checkLine' (di,dj) (i,j) r =
          do c <- readArray board (i,j)
             if c == ocolor then
                 checkLine' (di,dj) (i+di,j+dj) ( (i,j) : r )
             else if c == color then
                      return r
                  else
                      return []

{-
   boardをそのまま返すのは一般には危険。
   だが、今のdoMoveの使用方法ならば問題ない。
-}
doMove :: Board -> Mv -> Color -> IO Board
doMove board GiveUp  color = return board
doMove board Pass    color = return board
doMove board (M i j) color =
    do ms <- flippableIndices board color (i,j)
       mapM_ (\(ii,jj) -> writeArray board (ii,jj) color) ms
       writeArray board (i,j) color
       return board

noIOdoMove :: Board -> Mv -> Color -> IO Board
noIOdoMove board GiveUp  color = return board
noIOdoMove board Pass    color = return board
noIOdoMove board (M i j) color =
    do ms <- flippableIndices board color (i,j)
       board' <- foldrM (\(ii,jj) acc -> write acc (ii,jj) color) board ms
       write board' (i,j) color

write :: Board -> (Int, Int) -> Int -> IO Board
write board (i, j) color =
    do (l1, l2) <- splitAt (i * 10 + j) <$> getElems board
       newListArray ( (0, 0), (9, 9) ) $ l1 ++ (color : tail l2)

-- 合法手
validMoves :: Board -> Color -> IO [ (Int,Int) ]
validMoves board color =
     filterM (isValidMove board color)
             [ (i,j) | i <- [1..8], j <- [1..8]]

play :: Board -> Color -> IO Mv
play board color =
    do ms <- validMoves board color
       zeronum <- count board none
       case ms of
         [] -> return Pass
         _  | zeronum > 12 -> playFirst board color ms
            | zeronum <= 12 -> playLast board color ms

-- 序盤
playFirst :: Board -> Color -> [ (Int, Int) ] -> IO Mv
playFirst board color ms =
    case l1 of
      [] -> do l2 <- filterM (fixedStone board color) ms
               case l2 of
                 [] -> uncurry M . fst <$> foldrM (\e (e1, acc) ->
                         case acc of
                           Left x -> do aa <- ((*) (-1)) <$> (noIOdoMove board (uncurry M e) color >>= negaAlpha color 6 (-1000) (-x))
                                        if | aa > x -> return (e, Left aa)
                                           | otherwise -> return (e1, acc)
                           Right _ -> return (e1, acc)) ((0, 0), Left (-1000)) ms
                 _ -> do k <- getStdRandom $ randomR (0, length l2-1)  -- 確定石があるなら取る
                         return $ uncurry M (l2 !! k)
      _ -> do k <- getStdRandom $ randomR (0, length l1-1)   -- 角が取れるなら取る
              return $ uncurry M (l1 !! k)
 where l1 = filter (\(i, j) -> (i == 1 || i == 8) && (j == 1 || j == 8)) ms

-- 終盤 読み切り
playLast :: Board -> Color -> [ (Int, Int) ] -> IO Mv
playLast board color ms = uncurry M . fst <$> foldrM (\e (e1, acc) ->
                            case acc of
                              Left x -> do aa <- ((*) (-1)) <$> (noIOdoMove board (uncurry M e) color >>= negaAlpha2 color False (-1000) (-x))
                                           if | aa > x -> return (e, Left aa)
                                              | otherwise -> return (e1, acc)
                              Right _ -> return (e1, acc)) ((0, 0), Left (-1000)) ms

-- 石の位置による評価
boardEvalArray = listArray ((0, 0), (9, 9))
                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                 0, 30, -12, 0, -1, -1, 0, -12, 30, 0,
                 0, -12, -15, -3, -3, -3, -3, -15, -12, 0,
                 0, 0, -3, 0, -1, -1, 0, -3, 0, 0,
                 0, -1, -3, -1, -1, -1, -1, -3, -1, 0,
                 0, -1, -3, -1, -1, -1, -1, -3, -1, 0,
                 0, 0, -3, 0, -1, -1, 0, -3, 0, 0,
                 0, -12, -15, -3, -3, -3, -3, -15, -12, 0,
                 0, 30, -12, 0, -1, -1, 0, -12, 30, 0,
                 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

-- 評価関数
evaluate :: Board -> Color -> IO Int
evaluate board color =
    (+) <$> foldrM (\index acc ->
              do e <- readArray board index
                 case e of
                   _ | e == color -> return $ acc + (boardEvalArray ! index)
                     | e == oppositeColor color -> return $ acc - (boardEvalArray ! index)
                     | otherwise -> return acc)
        0 [ (i, j) | i <- [1 .. 8], j <- [1 .. 8] ]
       <*> validSpaceEval board color

-- 着手可能なマスの数
validSpaceEval :: Board -> Color -> IO Int
validSpaceEval board color =
    (-) <$> (length <$> validMoves board color)
        <*> (length <$> validMoves board (oppositeColor color))

-- 確定石
fixedStone :: Board -> Color -> (Int, Int) -> IO Bool
fixedStone board color (i, j)
    | (i == 1 || i == 8) && (j /= 1 && j /= 8) =
        (||) <$> allM (fmap (== color) . readArray board)
                     [(i, k) | k <- [1 .. j - 1]]
             <*> allM (fmap (== color) . readArray board)
                     [(i, k) | k <- [j + 1 .. 8]]
    | (j == 1 || j == 8) && (i /= 1 && i /= 8) =
        (||) <$> allM (fmap (== color) . readArray board)
                     [(k, j) | k <- [1 .. i - 1]]
             <*> allM (fmap (== color) . readArray board)
                     [(k, j) | k <- [i + 1 .. 8]]
    | (i == 2 || i == 7) && (j == 2 || j == 7) =
        (||) <$> allM (fmap (== color) . readArray board) (f1 (i, j))
             <*> allM (fmap (== color) . readArray board) (f2 (i, j))
    | otherwise = return False
 where f1 :: (Int, Int) -> [ (Int, Int) ]
       f1 (2, 2) = [(1, 1), (1, 2), (1, 3), (2, 1)]
       f1 (2, 7) = [(1, 8), (1, 7), (1, 6), (2, 8)]
       f1 (7, 2) = map swap $ f1 (2, 7)
       f1 (7, 7) = [(8, 8), (8, 7), (8, 6), (7, 8)]
       f2 :: (Int, Int) -> [ (Int, Int) ]
       f2 = map swap . f1

negaAlpha :: Color -> Int -> Int -> Int -> Board -> IO Int
negaAlpha color depth a b board
  | depth == 0 = if bool then evaluate board color else ((*) (-1)) <$> evaluate board color
  | otherwise = do 
      ms <- validMoves board nowcolor
      if | null ms -> negaAlpha color (depth-1) (-b) (-a) board
         | otherwise -> either id id <$> foldrM (\e acc -> case acc of
                                            Left x -> do aa <- (max x) . ((*) (-1)) <$> (noIOdoMove board (uncurry M e) nowcolor >>= negaAlpha color (depth-1) (-b) (-x))
                                                         if | aa >= b -> return $ Right aa
                                                            | otherwise -> return $ Left aa
                                            Right _ -> return acc) (Left a) ms
 where nowcolor = if bool then color else oppositeColor color -- 打つ人
       bool = depth `mod` 2 /= 0

negaAlpha2 :: Color -> Bool -> Int -> Int -> Board -> IO Int
negaAlpha2 color bool a b board =
    do isFin <- (== 0) <$> count board none
       ms <- validMoves board nowcolor
       ms' <- validMoves board (oppositeColor nowcolor)
       if | (isFin || (null ms && null ms')) -> if bool then count board color else ((*) (-1)) <$> count board color
          | null ms -> negaAlpha2 color (not bool) (-b) (-a) board
          | otherwise -> either id id <$> foldrM (\e acc -> case acc of
                                            Left x -> do aa <- (max x) . ((*) (-1)) <$> (noIOdoMove board (uncurry M e) nowcolor >>= negaAlpha2 color (not bool) (-b) (-x))
                                                         if | aa >= b -> return $ Right aa
                                                            | otherwise -> return $ Left aa
                                            Right _ -> return acc) (Left a) ms
 where nowcolor = if bool then color else oppositeColor color -- 打つ人

-- 石の数
count :: Board -> Color -> IO Int
count board color =
    do is <- filterM (\i ->
                          do e <- readArray board i
                             return $ e == color)
                     [ (i,j) | i <- [1..8], j <- [1..8]]
       return $ length is


-- 盤面の出力
putBoard :: Board -> IO ()
putBoard board =
    do putStrLn " |A B C D E F G H "
       putStrLn "-+----------------"
       mapM_ putBoardLine [1..8]
       putStrLn "  (X: Black,  O: White)"
    where
      putC c | c == none  = putStr " "
             | c == white = putStr "O"
             | c == black = putStr "X"
      putBoardLine j =
          do putStr $ show j ++ "|"
             mapM_ (\i -> do e <- readArray board (i,j)
                             putC e >> putStr " ") [1..8]
             putStrLn ""



Client.hs
{-# LANGUAGE BangPatterns #-}

module Main where

import System.Environment (getArgs) 
import Network 

import Command
import Play

import Text.Printf 
import System.IO

import Color 
import Control.Monad
import System.Console.GetOpt

data Config 
    = Config { host :: HostName,
               port :: PortNumber, 
               playerName :: String,
               verbose :: Bool,
               helpMode :: Bool } 

defaultConf = Config "localhost" 3000 "Anon." False False

options :: [OptDescr (Config -> Config)]
options =
    [ Option ['v'] ["verbose"]
             (NoArg $ \conf -> conf { verbose = True })
             "verbose mode"
    , Option ['H'] ["host"]
             (ReqArg (\s conf -> conf { host = s }) "HOST")
             "host name of a server" 
    , Option ['p'] ["port"]
             (ReqArg (\s conf -> conf { port = fromIntegral (read s :: Int) }) "PORT")
             "port number of a server"
    , Option ['n'] ["name"]
             (ReqArg (\s conf -> conf { playerName = s }) "NAME")
             "player name" 
    , Option ['h','?'] ["help"]
             (NoArg (\conf -> conf { helpMode = True }))
             "show this help"
    ]

usageMessage :: String
usageMessage = usageInfo header options 
    where
      header =
          "Usage: \n" 
          ++ "    reversi -H HOST -p PORT -n NAME ...\n" 

parseArg :: [String] -> IO (Config, [String]) 
parseArg args =
    case getOpt Permute options args of 
      (o, n, [])  -> return (foldl (flip ($)) defaultConf o, n)
      (_, _, err) -> ioError (userError (concat err ++ usageMessage))
      
hGetCommand :: Handle -> IO Command 
hGetCommand h = 
    do { r <- hGetLine h 
       ; putStrLn $ "Received: " ++ r 
       ; return $ either error id $ parseCommand r 
       }

hGetCommand' :: Handle -> IO Command
hGetCommand' h = 
    do { c <- hGetCommand h  
       ; case c of 
           Empty -> hGetCommand' h  
           _     -> return c }


    
hPutCommand :: Show a => Handle -> a -> IO () 
hPutCommand h c = 
    do hPutStr h (show c)
       hPutStr h "\n"
       hFlush h 
       putStrLn $ "Sent: " ++ show c 

-- History                         
data OPMove = OMove Mv | PMove Mv 

instance Show OPMove where 
    show (OMove mv) = "-" ++ show mv
    show (PMove mv) = "+" ++ show mv 
type Hist = [OPMove]

showHist :: [OPMove] -> String 
showHist hist = 
    foldr (\a r -> show a ++ " " ++ r) "" $ reverse hist 
    

showScores :: [(String, (Int, Int, Int))] -> String 
showScores scores =
    foldr (\(n,(s,w,l)) r -> 
               take (len+1) (n ++ ":" ++ cycle " ") 
                    ++ replicate (1 + slen - length (show s)) ' ' ++ show s 
                    ++ " (Win " ++ show w ++ ", Lose " ++ show l ++ ")\n" ++ r) "" scores 
    where
      len  = maximum $ map (length . fst) scores
      slen = maximum $ map (length . show . (\(a,b,c) -> a) . snd) scores 

-- クライアント
client :: Config -> IO () 
client !conf =
    do { putStrLn $ "Connecting to " ++ host conf ++ " " ++ show (port conf)
       ; h <- connectTo (host conf) (PortNumber $ port conf) 
       ; putStrLn "Connection Ok."
       ; hPutCommand h $ Open (playerName conf)
       ; waitStart h conf }

waitStart :: Handle -> Config -> IO () 
waitStart !h !conf =
    do { c <- hGetCommand' h 
       ; case c of 
           Bye scores -> 
               putStr $ showScores scores 
           Start color opname mytime ->
               do { board <- initBoard 
                  ; if color == black then 
                        performMyMove h board color [] opname mytime conf 
                    else
                        waitOpponentMove h board color [] opname mytime conf }
           _ -> 
               error $ "Invalid Command: " ++ show c 
       }

performMyMove :: Handle -> Board -> Color -> [OPMove] -> String -> Int -> Config -> IO ()
performMyMove h board color hist opname mytime conf =
    do { pmove <- play board color
       ; board <- doMove board pmove color
       ; hPutCommand h $ Move pmove 
       ; when (verbose conf) $ putStrLn $ replicate 80 '-'
       ; when (verbose conf) $ putStrLn ("PMove: " ++ show pmove ++ " " ++ showColor color) 
       ; putBoard board
       ; when (verbose conf) (putBoard board)
       ; c <- hGetCommand' h 
       ; case c of 
           Ack mytime' -> 
               waitOpponentMove h board color (PMove pmove:hist) opname mytime' conf 
           End wl n m r ->
               procEnd h board color hist opname wl n m r conf 
           _ -> 
               error $ "Invalid Command: " ++ show c 
       }

waitOpponentMove :: Handle -> Board -> Color -> [OPMove] -> String -> Int -> Config -> IO ()
waitOpponentMove h board color hist opname mytime conf =
    do { c <- hGetCommand' h
       ; case c of 
           Move omove ->
               do { board <- doMove board omove (oppositeColor color)
                  ; when (verbose conf) $ putStrLn $ replicate 80 '-'
                  ; when (verbose conf) $ putStrLn ("OMove: " ++ show omove ++ " " ++ showColor color) 
                  ; when (verbose conf) (putBoard board)
                  ; performMyMove h board color (OMove omove:hist) opname mytime conf }
           End wl n m r -> 
               procEnd h board color hist opname wl n m r conf 
           _ -> 
               error $ "Invalid Command: " ++ show c 
       }
           
procEnd :: Handle -> Board -> Color -> [OPMove] -> String -> WL -> Int -> Int -> String -> Config -> IO ()
procEnd h board color hist opname wl n m r conf =
    do { case wl of 
           Win ->
               putStrLn $ printf "You win! (%d vs. %d) -- %s." n m r
           Lose -> 
               putStrLn $ printf "You lose! (%d vs. %d) -- %s." n m r
           Tie ->
               putStrLn $ printf "Draw (%d vs. %d) -- %s." n m r
       ; putStrLn $ printf "Your name: %s (%s)  Oppnent name: %s (%s)."
                      (playerName conf) (showColor color)
                      opname (showColor (oppositeColor color))
       ; putBoard board 
       ; putStrLn $ showHist hist 
       ; waitStart h conf }
                  
main :: IO () 
main = withSocketsDo $ 
       do { args <- getArgs 
          ; (conf, rest) <- parseArg args 
          ; if helpMode conf then 
                putStrLn usageMessage 
            else 
                client conf }
          


Server.hs
{-# LANGUAGE BangPatterns #-}
module Main where 

import Network 
import System.Environment 
import System.Random 
import System.IO
import System.CPUTime

import Control.Monad
import Control.Concurrent
import Control.Concurrent.MVar 

import qualified Data.Map as M 

import Data.List (intercalate)
import Data.Time 
import Data.Maybe 
import Data.Tuple.Select
import Color 
import Command 
import Play 

import System.Console.GetOpt

data Config = 
    Config { port :: PortNumber
           , timeout :: Int -- in milliseconds 
           , playersTime :: Int -- in milliseconds
           , byoyomi :: Int 
           , numberOfPlayers :: Int 
           , numberOfRounds  :: Int 
           , verbose :: Int 
           , helpMode :: Bool 
           , concurrency :: Int 
           }

instance Show Config where
    show conf =
        unlines ["Server Configuration:"
                , "\tListening:           " ++ show (port conf) 
                , "\tTime out:            " ++ show (fromIntegral (timeout conf) / 1000) ++ " s" 
                , "\tEach Player's Time:  " ++ show (fromIntegral (playersTime conf) / 1000) ++ " s" 
                , "\tByoyomi:             " ++ show (byoyomi conf) ++ " ms"
                , "\t#Players:            " ++ show (numberOfPlayers conf) 
                , "\t#Rounds:             " ++ show (numberOfRounds conf) 
                , "\tVerbosity:           " ++ show (verbose conf) 
                , "\tDig. of Concurrency: " ++ show (concurrency conf) ++ " (effective if #Players >= 4)"]


defaultConf = Config 3000 (60 * 1000) (10 * 60 * 1000) 500 2 4 0 False 2

options :: [OptDescr (Config -> Config)]
options =
    [ Option ['v'] ["verbose"]
             (OptArg (\m conf -> maybe (conf { verbose = 1 }) (\s -> conf {verbose = read s}) m) "LEVEL")
             "verbose mode"
    , Option ['p'] ["port"]
             (ReqArg (\s conf -> conf { port = fromIntegral (read s :: Int) }) "PORT")
             "port number"
    , Option ['t'] ["playerstime"]
             (ReqArg (\s conf -> conf { playersTime = max (read s * 60 * 1000) 0 }) "TIME") 
             "time assigned to each player (in second)"
    , Option [] ["timeout"]
             (ReqArg (\s conf -> conf { timeout = read s }) "TIME")
             "time out (in millisecond)"
    , Option ['T'] ["byoyomi"]
             (ReqArg (\s conf -> conf { byoyomi = read s }) "TIME")
             "byoyomi time (in millisecond)"
    , Option ['n'] ["players"]
             (ReqArg (\s conf -> conf { numberOfPlayers = max (read s) 2 }) "NUM")
             "number of players"
    , Option ['h','?'] ["help"]
             (NoArg (\conf -> conf { helpMode = True }))
             "show this help"
    , Option ['r'] ["rounds"]
             (ReqArg (\s conf -> conf { numberOfRounds = (max (read s) 1 + 1) `div` 2 * 2 }) "NUM")
             "number of rounds (multiple of 2)"
    , Option ['c'] ["concurrency"]
             (ReqArg (\s conf -> conf { concurrency = max 1 (read s) }) "NUM")
             "number of maximum concurrent matches (EXPERIMENTAL)"
    ]

usageMessage :: String 
usageMessage = usageInfo header options 
    where
      header =
          "Usage: \n" 
          ++ "    reversi-serv -p PORT ...\n" 

parseArg :: [String] -> IO (Config, [String])
parseArg args =
    case getOpt Permute options args of 
      (o, n, [])  -> return (foldl (flip ($)) defaultConf o, n)
      (_, _, err) -> ioError (userError (concat err ++ usageMessage))


hGetCommand :: Handle -> Int -> Config -> IO (Maybe Command)
hGetCommand h t conf = 
    do { b <- hWaitForInput h t 
       ; if b then  
             do r <- hGetLine h 
                when (verbose conf >= 1) (putStrLn $ "Received: " ++ r)
                return $ either (const Nothing) Just $ parseCommand r 
         else
             return Nothing                 
       }

hGetCommand' :: Handle -> Int -> Config -> IO (Maybe Command) 
hGetCommand' h t conf = 
    do { c <- hGetCommand h t conf 
       ; case c of 
           Just Empty -> hGetCommand' h t conf 
           _          -> return c }

hPutCommand :: Handle -> Command -> Config -> IO ()
hPutCommand h c conf = 
    do hPutStr h (show c)
       hPutStr h "\n"
       hFlush h 
       when (verbose conf >= 1) (putStrLn $ "Sent: " ++ show c) 
                        

isValidMv Pass board color = 
    fmap null $ validMoves board color

isValidMv (M i j) board color =
    isValidMove board color (i,j) 


roundRobin []     = []
roundRobin (x:xs) = map (\y -> (x,y)) xs ++ roundRobin xs 

-- Naive Scheduling at most n parallel
greedySchedule :: Eq a => Int -> [(a,a)] -> [[(a,a)]]
greedySchedule n xs = go [] [] [] xs 
    where
      go rs p []   []         = [rs]
      go rs p rest []         = rs : go [] [] [] rest 
      go rs p rest ((x,y):xs) 
          | length rs >= n = 
              rs : go [(x,y)] [x,y] [] (rest++xs)
          | x `elem` p || y `elem` p = 
              go rs p ((x,y):rest) xs 
          | otherwise = 
              go ((x,y):rs) (x:y:p) rest xs

type PlayerName  = String 
type PlayerInfo  = (Handle, PlayerName, HostName, PortNumber)

type Score = Int
type WinCount  = Int
type LoseCount = Int 

type Result = M.Map PlayerName (Score, WinCount, LoseCount)

winScore :: Score
winScore  =  1 

loseScore :: Score 
loseScore = -1

tieScore :: Score 
tieScore  =  0 

add3 (a,b,c) (x,y,z) = (a+x,b+y,c+z) 

server :: Config -> IO () 
server conf = 
    do socket <- listenOn (PortNumber (port conf))
       loop socket 
    where
      putWaitingMsg :: Int -> IO () 
      putWaitingMsg n 
          | n == 0 = return ()
          | n == 1 =
              putStrLn "Waiting 1 connection ..."
          | otherwise = 
              putStrLn $ "Waiting " ++ show n ++ " connections..."

      waitPlayers :: Socket -> Int -> IO [PlayerInfo]
      waitPlayers socket 0 = return []
      waitPlayers socket n = 
          do (ph1, cHost1, cPort1) <- accept socket 
             c <- hGetCommand' ph1 (timeout conf) conf 
             case c of 
                 Just (Open name) -> 
                     do putWaitingMsg (n-1)
                        plist <- waitPlayers socket (n-1)
                        return $ (ph1,name,cHost1,cPort1):plist 
                 _ -> 
                     waitPlayers socket n 

      loop :: Socket -> IO () 
      loop !socket =
          do putWaitingMsg (numberOfPlayers conf) 
             players <- waitPlayers socket (numberOfPlayers conf) 
             forkIO (startSessions players) 
             loop socket 

      -- 総当たり戦の実行
      startSessions :: [PlayerInfo] -> IO () 
      startSessions players =
          do let playerNames = map (\(_,n,_,_) -> n) players 
             putStrLnWithTime $ "Session started (" ++ intercalate ", " playerNames ++ ")."
             let matches = roundRobin players 
             let mss = greedySchedule (concurrency conf) matches 
             res <- concurrentMatch mss 
             mapM_ (\(h,n, _, _) -> 
                         do hPutCommand h (Bye $ M.toList res) conf 
                            hClose h) players 
             putStrLnWithTime $ "Session finished (" ++ intercalate ", "  playerNames ++ ")."
             

      putStrLnWithTime s = 
          do lt <- getZonedTime 
             putStrLn $ "[" ++ show lt ++ "] " ++ s 


      -- 並行に対戦
      concurrentMatch :: [[(PlayerInfo,PlayerInfo)]] -> IO Result 
      concurrentMatch []       = return M.empty 
      concurrentMatch (ms:mss) = 
          do mvs  <- mapM (const newEmptyMVar) ms
             rref <- newMVar M.empty 
             mapM_ (\(mvar,m) -> forkFinally (oneStep rref m)
                                             (\_ -> putMVar mvar ())) $ zip mvs ms 
             mapM_ takeMVar mvs  -- wait for child processes  
             r1 <- takeMVar rref 
             r2 <- concurrentMatch mss 
             return $ M.unionWith add3 r1 r2 
          where
            oneStep rref m = 
                do res <- startMatch m 
                   modifyMVar_ rref (return . M.unionWith add3 res)


      -- 対戦を先手後手を入れ替えつつcount回行う
      startMatch :: (PlayerInfo,  PlayerInfo) -> IO Result
      startMatch ((ph1, name1, _, _), (ph2, name2, _, _)) =
          do putStrLnWithTime $ name1 ++ " vs. " ++ name2 ++ " started. "
             k <- getStdRandom $ randomR (0, 1)
             let [(h1,n1),(h2,n2)] = 
                       map ([(ph1,name1), (ph2,name2)]!!) [k,1-k]
             startMatches (n1,h1) (n2,h2) (numberOfRounds conf) 

      -- 上の補助関数
      startMatches :: (PlayerName,Handle) -> (PlayerName,Handle) -> Int -> IO Result
      startMatches (n1,h1) (n2,h2) count =
          if count > 0 then 
              do board <- initBoard 
                 let t = playersTime conf 
                 hPutCommand h1 (Start black n2 t) conf 
                 hPutCommand h2 (Start white n1 t) conf 
                 rr1 <- interact (n1,h1,t) (n2,h2,t) black board False 
                 -- let xs = rr1 M.! "kotapiku"
                 -- if sel2 xs == 1 then putStrLn $ "round " ++ show (numberOfRounds conf - count + 1) ++ " : win!"
                 --                 else putStrLn $ "round " ++ show (numberOfRounds conf - count + 1) ++ " : lose..."
                 rr2 <- startMatches (n2,h2) (n1,h1) (count - 1) 
                 return $ M.unionWith add3 rr1 rr2 
          else
              return M.empty 
              

      -- 実際の対戦処理
      interact :: (PlayerName,Handle,Int) -> (PlayerName,Handle,Int) -> Color -> Board -> Bool -> IO Result
      interact (n1,h1,t1) (n2,h2,t2) color board passFlag =
          do { startTime <- getCurrentTime 
             ; command   <- hGetCommand' h1 (timeout conf) conf 
             ; endTime   <- getCurrentTime 
             ; let diff = round (1000 * realToFrac(endTime `diffUTCTime` startTime)) 
             ; when (verbose conf >= 1) (putStrLn $ "Passed: " ++ show diff ++ " ms")
             ; let t1' = t1 - diff 
             ; if (t1' <= 0) && (diff > byoyomi conf) then 
                   suddenEnd (n1,h1,t1) (n2,h2,t2) color board "TIME_UP"
               else
                   case command of 
                     Just (Move m) ->
                         do { isValid <- isValidMv m board color 
                            ; if not isValid then do
                                  suddenEnd (n1,h1,t1) (n2,h2,t2) color board "ILLEGAL_MOVE"
                              else
                                  case m of 
                                    Pass | passFlag -> 
                                         do { count1 <- count board color
                                            ; count2 <- count board (op color)
                                            ; let (r1,r2,rr) 
                                                   | count1 == count2 = 
                                                       (Tie,Tie, M.fromList [(n1,tieResult),(n2,tieResult)])
                                                   | count1 >  count2 = 
                                                       (Win,Lose,M.fromList [(n1,winResult),(n2,loseResult)])
                                                   | count1 <  count2 = 
                                                       (Lose,Win,M.fromList [(n1,loseResult),(n2,winResult)])
                                            ; hPutCommand h1 (End r1 count1 count2 "DOUBLE_PASS") conf 
                                            ; hPutCommand h2 (End r2 count2 count1 "DOUBLE_PASS") conf 
                                            ; return rr }
                                    _ -> 
                                        do { board <- doMove board m color
                                           ; hPutCommand h1 (Ack t1') conf 
                                           ; hPutCommand h2 (Move m) conf 
                                           ; interact (n2,h2,t2) (n1,h1,t1') (op color) board (case m of { Pass -> True; _ -> False }) }}
                     
                     _ -> 
                         suddenEnd (n1,h1,t1) (n2,h2,t2) color board "ILLEGAL_COMMAND"
             }

      -- 変なコマンドが送られてきたときの処理
      suddenEnd (n1,h1,t1) (n2,h2,t2) color board desc =
          do count1 <- count board color 
             count2 <- count board (op color) 
             hPutCommand h1 (End Lose count1 count2 desc) conf
             hPutCommand h2 (End Win  count2 count1 desc) conf 
             return $ M.fromList [ (n1, loseResult), (n2, winResult) ]
          
      op = oppositeColor 

      winResult  = (winScore,1,0)
      loseResult = (loseScore,0,1)
      tieResult  = (tieScore,0,0)
 
      -- parallelMatchmodify mss = 
      --     do { counter <- newMVar 0 
      --        ; go counter mss }
      --     where
      --       go counter [] = return M.empty 
      --       go counter (ms:mss) =
      --           do { modifyMVar_ counter (\_ -> return $ length ms)
      --              ; rref <- newMVar M.empty 
      --              ; mapM_ (oneStep counter rref) ms
      --              ; r1 <- checkAfterZero counter (readMVar rref) 
      --              ; r2 <- go counter mss 
      --              ; return $ M.unionWith (+) r1 r2 }
      --           where
      --             checkAfterZero counter action =
      --                 do { c <- readMVar counter 
      --                    ; if c == 0 then 
      --                          action
      --                      else 
      --                          do { threadDelay (100 * 1000) 
      --                             ; checkAfterZero counter action }}
      --             oneStep counter rref m = forkIO $ 
      --                do { res <- startMatch m 
      --                   ; modifyMVar_ rref (return . M.unionWith (+) res)
      --                   ; modifyMVar_ counter (\x -> return $ x - 1) }
                                                                  


main :: IO ()                                 
main = withSocketsDo $ 
       do args <- getArgs 
          (conf,rest) <- parseArg args
          if helpMode conf then 
              putStrLn usageMessage 
          else
              do print conf 
                 server conf
